keps/sig-node/runtime-class-scheduling.md

```diff
@@ -0,0 +1,347 @@
+tolerations with the existing tolerations, which matches the default toleration
+composition logic. This means that RuntimeClass tolerations can simply be
+appended to the existing tolerations, but an [existing
+utilty][merge-tolerations] can reduce duplicates by merging equivalent
+tolerations.
+
+[merge-tolerations]: https://github.com/kubernetes/kubernetes/blob/58021216b16ae6d5f24fb1c32ab541b2e79a365e/pkg/util/tolerations/tolerations.go#L62
+
+### Labeling Nodes
+
+Node labeling & tainting is outside the scope of this proposal or feature. How
+to label nodes is very environment depnedent. Here are several examples:
+
+- If runtimes are setup as part of node setup, then the node template should
+  include the appropriate labels & taints.
+- If runtimes are installed through a DaemonSet, then the topology should match
+  that of the DaemonSet.
+- If runtimes are manually installed, or installed through some external
+  process, that same process should apply an appropriate label to the node.
+
+### Graduation Criteria
+
+This feature will be rolled into RuntimeClass beta in v1.15, thereby skipping
+the alpha phase. This means the feature is expected to be beta quality at launch:
+
+- Thorough testing, including unit, integration and e2e coverage.
+- Thoroughly documented (as an extension to the [RuntimeClass documentation][]).
+
+[RuntimeClass documentation]: https://kubernetes.io/docs/concepts/containers/runtime-class/
+
+## Implementation History
+
+- 2019-03-14: Initial KEP published.
+- 2018-10-05: [RuntimeClass Scheduling
+  Brainstorm](https://docs.google.com/document/d/1W51yBNTvp0taeEss56GTk8jczqFJ2d6jBeN6sCSlYZU/edit#)
+  published.
+
+## Alternatives
+
+### NodeSelector
+
+Replacing the NodeSelector's `[]NodeSelectorRequirements` type with the
+PodSpec's label `map[string]string` approach greatly simplifies the merging
+logic, but sacrifices a lot of flexibliity. For exameple, the operator in
+NodeSelectorRequriments enables selections like:
+
+- Negative selection, such as "operating system is _not_ windows"
+- Numerical comparison, such as "runc version is _at least_ X" (although it doesn't currently support semver)
+- Set selection, such as "sandbox is _one of_ kata-cotainers or gvisor"
+
+### RuntimeClass-aware scheduling
+
+Rather than the proposed mix-in approach, the scheduler could have built in
+support for RuntimeClasses. When a pod is to be scheduled, the scheduler plugin
+would fetch the pod's RuntimeClass and determine which nodes are supported.
+
+**Advantages:**
+
+- Native scheduler support is required if we want to intersect more advanced
+  scheduling constraints, such as NodeAffinities.
+- Native scheduler support is required if we ever want a feedback loop between
+  the scheduling decision and the RuntimeClass selection. That is, if we want to
+  select a RuntimeClass based on where there is availibility.
+- Doesn't "pollute" the PodSpec with the RuntimeClass scheduling constraints.
```

Comment:
There is also a user experience advantage. If we had a RuntimeClass predicate, the scheduler would throw unschedulable messages like: 0/10 Nodes are availeble: 5 nodes do not have enough memory, 5 nodes didn't match RuntimeClass.

With the proposed solution, a user will see an unschedulable message like: 
0/10 Nodes are availeble: 5 nodes do not have enough memory, 5 nodes didn't match NodeAffinity.

If a user didn't specify any NodeAffinity they may get confused why their pods are not scheduled. Worse is that if they specify NodeAffinity, they may think that their NodeAffinity prevented pods from getting scheduled. So, they may start changing their affinity rules, not knowing that the affinity rules are added automatically by the RuntimeClass admission controller.

Reply:
Interesting point. How do you think tolerations would fit into this? Would you expect the tolerations to still be mixed in to the pod spec, or for `PodToleratesNodeTaints` predicate to fetch the tolerations from both the pod & the runtime class?

Reply:
After giving it more thoughts I think I agree with @bsalamat on the UX point, especially on scenarios where the operator of the cluster isn't the end user deploying into it (which is a rather common use case for multi-tenant clusters). 

Basically this would boil down to the scheduling watching `RuntimeClass`, build a cache out of it and use it on a runtimeClassPredicate that would be a composition of PodToleratesNodeTaints and affinity ones.

But I also still think that we can switch to a predicate if we see a long term benefit

Reply:
Let's move the conversation to here: https://github.com/kubernetes/enhancements/pull/909

I think I agree. I'm still a little unsure about the tolerations though, and whether those should remain a mixin.

Reply:
Regarding tolerations, I think it depends on how we tag the nodes - via (1) node label pairs only (2) with taints only or (3) with both labels and taints.